### PR TITLE
[Bug] Fixes self-sacrifice moves (explosion, self-destruct) to prevent post-faint ability triggers (aftermath, innards out)

### PR DESCRIPTION
**Pull request by schmidtc1**
Rebranch of #2774 

## What are the changes?

Solves the issue of #2621 
Post faint ability attributes will trigger after a self-sacrifice move on the opposing pokemon.
With Aftermath, this occurs if the opposing pokemon used a physical move on the self-sacrificing pokemon prior to sacrifice.
With Innards Out, this occurs this occurs if the opposing pokemon used either a physical or special move on the self-sacrificing pokemon prior to sacrifice.

## Why am I doing these changes?

Post faint abilities (aside from PostFaintClearWeather abilities) should not trigger against the opposing pokemon when it uses a sacrificial move on itself.

## What did change?

Sacrifice moves are added to the attacksReceived history and a condition is placed on applyPostFaintAbAttrs to prevent PostFaint abilities (that are not PostFaintClearWeather abilities) from triggering when a sacrifice move is used.

### Screenshots/Videos

Before fix (Aftermath procs)

https://github.com/pagefaultgames/pokerogue/assets/62030095/ea598612-7991-4fd5-a8e9-3f9a7edac34f

After fix (Aftermath doesn't proc)

https://github.com/pagefaultgames/pokerogue/assets/62030095/86b9f239-2827-42b4-a79b-16b1328fb702

After fix (Desolate Land clears weather on faint)

https://github.com/pagefaultgames/pokerogue/assets/62030095/e16cb30c-35d5-4611-bfb0-27a8133ba026

## How to test the changes?

Give opponent self-sacrificing moves, hit opponent with a faster mon (if Aftermath is used, it must be a physical move). Observe the ability no longer triggers against opposing pokemon when the self-sacrifice move completes.

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [ ] Have I considered writing automated tests for the issue?
- [ ] Have I tested the changes (manually)?
    - [ ] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/phases.ts [hunk 1]
```diff
@@ -3795,7 +3795,9 @@ export class FaintPhase extends PokemonPhase {
 
     if (pokemon.turnData?.attacksReceived?.length) {
       const lastAttack = pokemon.turnData.attacksReceived[0];
-      applyPostFaintAbAttrs(PostFaintAbAttr, pokemon, this.scene.getPokemonById(lastAttack.sourceId), new PokemonMove(lastAttack.move).getMove(), lastAttack.result);
+      if (pokemon.hasAbilityWithAttr(PostFaintClearWeatherAbAttr) || lastAttack.sourceId !== pokemon.id) { // prevent applying faint abilities to self unless its clear weather
+        applyPostFaintAbAttrs(PostFaintAbAttr, pokemon, this.scene.getPokemonById(lastAttack.sourceId), new PokemonMove(lastAttack.move).getMove(), lastAttack.result);
```
Comment (DayKev) [inline src/phases.ts:3799]: What if a Primal Kyogre is fused with a Pokémon with a `PostFaintAbAttr` as part of its passive?

Comment (schmidtc1) [reply src/phases.ts]: Solid edge case, testing with Innards Out passive fused with Primal Kyogre causes both abilities to trigger on faint from Self-Destruct.
I'll try to solve this.

**Comment by frutescens** [comment]
Hi, have you taken a look at PR #3492 's implementation of PostFaintAbAttr ?

**Comment by schmidtc1** [comment]
> Hi, have you taken a look at PR #3492 's implementation of PostFaintAbAttr ?

Looks like that implementation has significant overlap with this one, but I don't think it would resolve the issue of sacrificial moves like Explosion and Self-Destruct triggering PostFaintAbAttrs.

File: src/data/move.ts [hunk 2]
```diff
@@ -1419,7 +1423,10 @@ export class HalfSacrificialAttr extends MoveEffectAttr {
     // Check to see if the Pokemon has an ability that blocks non-direct damage
     applyAbAttrs(BlockNonDirectDamageAbAttr, user, cancelled);
     if (!cancelled.value) {
-      user.damageAndUpdate(Utils.toDmgValue(user.getMaxHp()/2), HitResult.OTHER, false, true, true);
+      const damage = Utils.toDmgValue(user.getMaxHp()/2);
```
Comment (DayKev) [inline src/data/move.ts:1426]:
```suggestion
      const damage = Utils.toDmgValue(user.getMaxHp() / 2);
```

**Comment by DayKev** [comment]
Please add tests (and fix merge conflict).

**Pull request closed without merging**